zcash_client_backend/src/fees.rs

```diff
@@ -0,0 +1,163 @@
+use zcash_primitives::{
+    consensus::{self, BlockHeight},
+    transaction::{
+        components::{
+            amount::{Amount, BalanceError},
+            sapling::builder::{SaplingInput, SaplingOutput},
+            transparent::{builder::TransparentInput, TxOut},
+        },
+        fees::{FeeRule, FixedFeeRule},
+    },
+};
+
+/// A proposed change amount and output pool.
+pub enum ChangeValue {
+    Sapling(Amount),
+}
+
+impl ChangeValue {
+    pub fn value(&self) -> Amount {
+        match self {
+            ChangeValue::Sapling(value) => *value,
+        }
+    }
+}
+
+/// The amount of change and fees required to make a transaction's inputs and
+/// outputs balance under a specific fee rule, as computed by a particular
+/// [`ChangeStrategy`] that is aware of that rule.
+pub struct TransactionBalance {
+    proposed_change: Vec<ChangeValue>,
+    fee_required: Amount,
+}
+
+impl TransactionBalance {
+    /// Constructs a new balance from its constituent parts.
+    pub fn new(proposed_change: Vec<ChangeValue>, fee_required: Amount) -> Self {
+        TransactionBalance {
+            proposed_change,
+            fee_required,
+        }
+    }
+
+    /// The change values proposed by the [`ChangeStrategy`] that computed this balance.  
+    pub fn proposed_change(&self) -> &[ChangeValue] {
+        &self.proposed_change
+    }
+
+    /// Returns the fee computed for the transaction, assuming that the suggested
+    /// change outputs are added to the transaction.
+    pub fn fee_required(&self) -> Amount {
+        self.fee_required
+    }
+}
+
+/// Errors that can occur in balance
+pub enum ChangeError<E> {
+    InsufficientFunds { available: Amount, required: Amount },
+    StrategyError(E),
+}
+
+/// A trait that represents the ability to compute the suggested change and fees that must be paid
+/// by a transaction having a specified set of inputs and outputs.
+pub trait ChangeStrategy {
+    type FeeRule: FeeRule;
+    type Error;
+
+    /// Returns the fee rule that this change strategy will respect when performing
+    /// balance computations.
+    fn fee_rule(&self) -> Self::FeeRule;
+
+    /// Computes the totals of inputs, suggested change amounts, and fees given the
+    /// provided inputs and outputs being used to construct a transaction.
+    ///
+    /// The fee computed as part of this operation should take into account the prospective
+    /// change outputs recommended by this operation. If insufficient funds are available to
+    /// supply the requested outputs and required fees, implementations should return
+    /// [`ChangeError::InsufficientFunds`].
+    fn compute_balance<P: consensus::Parameters>(
+        &self,
+        params: &P,
+        target_height: BlockHeight,
+        transparent_inputs: &[impl TransparentInput],
+        transparent_outputs: &[TxOut],
+        sapling_inputs: &[impl SaplingInput],
+        sapling_outputs: &[SaplingOutput],
```

Comment:
The benefit of adding an `impl SaplingOutput` trait would be additional information hiding from the caller; I'm not convinced that such information hiding is particularly useful; the `TransparentInput` and `SaplingInput` traits are necessary because the structs that they expose information from are private to their modules.